File: 501. 二叉搜索树中的众数/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def traversal(self, root, tlist):
        if not root:
            return
        self.traversal(root.left, tlist)
        tlist.append(root.val)
        self.traversal(root.right, tlist)

    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        if not root.left and not root.right:
            return [root.val]
        tlist = []
        ans = []
        self.traversal(root, tlist)
        Maxcount = 1
        count = 1
        pre = tlist[0]
        for i in range(1, len(tlist)):
            if pre == tlist[i]:
                count += 1
            if pre != tlist[i] or i == len(tlist)-1:
                if count > Maxcount:
                    ans.clear()
                    ans.append(pre)
                    Maxcount = count
                elif count == Maxcount:
                    ans.append(pre)
                pre = tlist[i]
                count = 1
            if i == len(tlist)-1:
                if count > Maxcount:
                    ans.clear()
                    ans.append(pre)
                    Maxcount = count
                elif count == Maxcount:
                    ans.append(pre)
        return ans

File: 501. 二叉搜索树中的众数/Solution.cpp
#include<bits/stdc++.h>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int maxCount = 0; // 最大频率
    int count = 0; // 统计频率
    TreeNode* pre = NULL;
    vector<int> result;
    void searchBST(TreeNode* cur) {
        if (cur == NULL) return ;

        searchBST(cur->left);       // 左
                                    // 中
        if (pre == NULL) { // 第一个节点
            count = 1;
        } else if (pre->val == cur->val) { // 与前一个节点数值相同
            count++;
        } else { // 与前一个节点数值不同
            count = 1;
        }
        pre = cur; // 更新上一个节点

        if (count == maxCount) { // 如果和最大值相同，放进result中
            result.push_back(cur->val);
        }

        if (count > maxCount) { // 如果计数大于最大值频率
            maxCount = count;   // 更新最大频率
            result.clear();     // 很关键的一步，不要忘记清空result，之前result里的元素都失效了
            result.push_back(cur->val);
        }

        searchBST(cur->right);      // 右
        return ;
    }

public:
    vector<int> findMode(TreeNode* root) {
        count = 0;
        maxCount = 0;
        pre = NULL; // 记录前一个节点
        result.clear();

        searchBST(root);
        return result;
    }
};
